iroh-sync/src/store/fs.rs

```diff
@@ -237,6 +250,95 @@ impl super::Store for Store {
+                    // super::PEERS_PER_DOC_CACHE_SIZE is non zero
+                    drop(namespace_peers);
+                    peers_table.insert(namespace, (nanos, peer))?;
+                }
+                Some((oldest_nanos, oldest_peer)) => {
+                    let oldest_peer = &oldest_peer;
+
+                    if oldest_peer == peer {
+                        // oldest peer is the current one, so replacing the entry for the peer will
+                        // maintain the size
+                        drop(namespace_peers);
+                        peers_table.remove(namespace, (oldest_nanos, oldest_peer))?;
+                        peers_table.insert(namespace, (nanos, peer))?;
+                    } else {
+                        // calculate the len in the same loop since calling `len` is another fallible operation
+                        let mut len = 1;
+                        // find any previous entry for the same peer to remove it
+                        let mut prev_peer_nanos = None;
+
+                        for result in namespace_peers {
+                            len += 1;
+                            let guard = result?;
+                            let (peer_nanos, peer_bytes) = guard.value();
+                            if prev_peer_nanos.is_none() && peer_bytes == peer {
+                                prev_peer_nanos = Some(peer_nanos)
+                            }
+                        }
+
+                        match prev_peer_nanos {
+                            Some(prev_nanos) => {
+                                // the peer was already present, so we can remove the old entry and
+                                // insert the new one without checking the size
+                                peers_table.remove(namespace, (prev_nanos, peer))?;
+                                peers_table.insert(namespace, (nanos, peer))?;
+                            }
+                            None => {
+                                // the peer is new and the table is non empty, add it and check the
+                                // size to decide if the oldest peer should be evicted
+                                peers_table.insert(namespace, (nanos, peer))?;
+                                len += 1;
+                                if len > super::PEERS_PER_DOC_CACHE_SIZE.get() {
+                                    peers_table.remove(namespace, (oldest_nanos, oldest_peer))?;
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        write_tx.commit()?;
+
+        Ok(())
+    }
+
+    fn get_sync_peers(&self, namespace: &NamespaceId) -> Result<Option<Self::PeersIter<'_>>> {
+        let read_tx = self.db.begin_read()?;
+        let peers_table = read_tx.open_multimap_table(NAMESPACE_PEERS_TABLE)?;
+        let mut peers = Vec::with_capacity(super::PEERS_PER_DOC_CACHE_SIZE.get());
```

Comment:
I tried to check this based on the other todos but I'm worried this will keep some underlying lock around for as long as the response is kept. Not sure, what do you think?

Reply:
well it would just keep the read tx around, so that should be fine, or at least the caller can decide if immediately to colllect or not